be/src/olap/rowset/segment_v2/inverted_index_query.cpp

```diff
@@ -0,0 +1,284 @@
+    if (_high_value > _low_value) {
+        switch (op) {
+        case InvertedIndexQueryOp::GREATER_THAN_QUERY: {
+            if (value >= _low_value) {
+                _low_value = value;
+                //_low_value_str = value_str;
+                // NOTE:full_encode_ascending will append encoded data to the end of buffer.
+                // so we need to clear buffer first.
+                _low_value_encoded.clear();
+                _value_key_coder->full_encode_ascending(&value, &_low_value_encoded);
+                _low_op = op;
+            }
+            break;
+        }
+
+        case InvertedIndexQueryOp::GREATER_EQUAL_QUERY: {
+            if (value > _low_value) {
+                _low_value = value;
+                //_low_value_str = value_str;
+                _low_value_encoded.clear();
+                _value_key_coder->full_encode_ascending(&value, &_low_value_encoded);
+                _low_op = op;
+            }
+            break;
+        }
+
+        case InvertedIndexQueryOp::LESS_THAN_QUERY: {
+            if (value <= _high_value) {
+                _high_value = value;
+                //_high_value_str = value_str;
+                _high_value_encoded.clear();
+                _value_key_coder->full_encode_ascending(&value, &_high_value_encoded);
+                _high_op = op;
+            }
+            break;
+        }
+
+        case InvertedIndexQueryOp::LESS_EQUAL_QUERY: {
+            if (value < _high_value) {
+                _high_value = value;
+                //_high_value_str = value_str;
+                _high_value_encoded.clear();
+                _value_key_coder->full_encode_ascending(&value, &_high_value_encoded);
+                _high_op = op;
+            }
+            break;
+        }
+
+        default: {
+            return Status::InternalError(
+                    "Add value string fail! Unsupported InvertedIndexQueryOp {}", op);
+        }
+        }
+    }
+
+    return Status::OK();
+}
+
+template <>
+bool InvertedIndexQuery<PrimitiveType::TYPE_VARCHAR>::has_lower_bound() {
+    return lower_value().data != &StringRef::MIN_CHAR;
+}
+
+template <>
+bool InvertedIndexQuery<PrimitiveType::TYPE_CHAR>::has_lower_bound() {
+    return lower_value().data != &StringRef::MIN_CHAR;
+}
+
+template <>
+bool InvertedIndexQuery<PrimitiveType::TYPE_STRING>::has_lower_bound() {
+    return lower_value().data != &StringRef::MIN_CHAR;
+}
+
+template <>
+bool InvertedIndexQuery<PrimitiveType::TYPE_VARCHAR>::has_upper_bound() {
+    return upper_value().data != &StringRef::MAX_CHAR;
+}
+
+template <>
+bool InvertedIndexQuery<PrimitiveType::TYPE_CHAR>::has_upper_bound() {
+    return upper_value().data != &StringRef::MAX_CHAR;
+}
+
+template <>
+bool InvertedIndexQuery<PrimitiveType::TYPE_STRING>::has_upper_bound() {
+    return upper_value().data != &StringRef::MAX_CHAR;
+}
+
+template <>
+Status InvertedIndexQuery<PrimitiveType::TYPE_VARCHAR>::add_fixed_value(InvertedIndexQueryOp op,
```

Comment:
In the header file, has already define and implement the function `add_fixed_value`, why need to implement `TYPE_VARCHAR`,`TYPE_CHAR`,`TYPE_STRING` these three types functions in the cpp file specially?

Reply:
because they do not need to do full_encode_ascending

Reply:
but i can refine the code.